File: couchbase-entity/src/main/java/com/kaufland/generation/CblConstantGeneration.java
package com.kaufland.generation;

import com.kaufland.model.entity.BaseEntityHolder;
import com.kaufland.model.field.CblBaseFieldHolder;
import com.kaufland.model.field.CblFieldHolder;
import com.kaufland.util.TypeUtil;
import com.squareup.javapoet.CodeBlock;
import com.squareup.javapoet.MethodSpec;

import javax.lang.model.element.Modifier;
import javax.lang.model.type.TypeMirror;

public class CblConstantGeneration {

    public static MethodSpec addConstants(BaseEntityHolder holder){
        MethodSpec.Builder builder = MethodSpec.methodBuilder("addConstants").
                addModifiers(Modifier.PRIVATE).
                addParameter(TypeUtil.createMapStringObject(), "map");

        for (CblBaseFieldHolder fieldHolder : holder.getFieldConstants()) {

            if (fieldHolder.isConstant()) {
                builder.addStatement("map.put($N, " + getConvertedValue(fieldHolder.getTypeMirror(), fieldHolder.getDefaultValue())+")", fieldHolder.getConstantName());
            }
        }
        return builder.build();
    }

    public static CodeBlock addAddCall(String nameOfMap){
       return CodeBlock.builder().addStatement("addConstants($N)", nameOfMap).build();
    }

    private static String getConvertedValue(TypeMirror clazz, String value) {

        if (clazz.toString().equals(String.class.getCanonicalName())) {
            return "\"" + value + "\"";
        }
        return value;
    }

}

File: couchbase-entity/src/main/java/com/kaufland/generation/CblConstantGeneration.kt
package com.kaufland.generation

import com.kaufland.model.entity.BaseEntityHolder
import com.kaufland.util.TypeUtil
import com.squareup.kotlinpoet.CodeBlock
import com.squareup.kotlinpoet.FunSpec
import com.squareup.kotlinpoet.KModifier
import com.squareup.kotlinpoet.asTypeName
import javax.lang.model.type.TypeMirror

object CblConstantGeneration {

    fun addConstants(holder: BaseEntityHolder): FunSpec {
        val builder = FunSpec.builder("addConstants").addModifiers(KModifier.PRIVATE).addParameter("map", TypeUtil.mutableMapStringObject())

        for (fieldHolder in holder.fieldConstants) {

            if (fieldHolder.isConstant) {
                builder.addStatement("map.put(%N, DOC_%N)", fieldHolder.constantName, fieldHolder.constantName)
            }
        }
        return builder.build()
    }

    fun addAddCall(nameOfMap: String): CodeBlock {
        return CodeBlock.builder().addStatement("addConstants(%N)", nameOfMap).build()
    }

}
